### PR TITLE
Use aligned version of operator new instead of aligned_malloc

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/6367. ~~Apparently, `aligned_malloc` is not available on Mac OS X before 10.15 and I don't see much of a problem using `posix_memalign` for Mac OS X unconditionally until we don't support 10.15 anymore.~~
The aligned version of operator new seems to work better (icpc needs some non-standard include file, though.).

### CODE REVIEW

File: core/src/impl/Kokkos_HostSpace.cpp [hunk 1]
```diff
@@ -91,6 +91,9 @@ void *HostSpace::impl_allocate(
     size_t size = (arg_alloc_size + alignment - 1) / alignment * alignment;
 #ifdef KOKKOS_COMPILER_MSVC
     ptr = _aligned_malloc(size, alignment);
+#elif defined(__APPLE__)
+    // FIXME_APPLE MacOS X doesn't provide std::aligned_alloc before 10.15
+    posix_memalign(&ptr, alignment, size);
```
Comment (jwakely) [inline core/src/impl/Kokkos_HostSpace.cpp:96]: `posix_memalign` is undefined unless `alignment >= sizeof(void*)`. Is that a precondition of this function already, or do you need to increase the alignment if it's small?

Comment (dalg24) [reply core/src/impl/Kokkos_HostSpace.cpp:96]: The alignment is always 64 as far as I can tell https://github.com/kokkos/kokkos/blob/ca3687e637c12225e824ee626e69fc9fea8cebd2/core/src/Kokkos_Macros.hpp#L190-L192

Looks like `alignment`
> must be a power of two and a multiple `of sizeof(void *)`

So I suppose we're good.

Comment (masterleinad) [reply core/src/impl/Kokkos_HostSpace.cpp:100]: Unless it's user-provided as compile definition (which probably almost never happens). I added a `static_assert` so we are safe either way.

**Review by dalg24 — DISMISSED** [review]
You forgot the deallocation

**Comment by dalg24** [comment]
Why did you chose this approach rather than the one suggested by Jonathan https://github.com/kokkos/kokkos/issues/6367#issuecomment-1683435438

```C++
#if __cpp_aligned_new >= 201606
  ptr = new (std::align_val_t(alignment)) char[size];
#else
  // fallbacks

**Comment by masterleinad** [comment]
> You forgot the deallocation

Can you please elaborate? `posix_memalign` allocations are also deallocated using `free`.

**Comment by masterleinad** [comment]
> Why did you chose this approach rather than the one suggested by Jonathan [#6367 (comment)](https://github.com/kokkos/kokkos/issues/6367#issuecomment-1683435438)

It's not that clear to me what the fallback(s) would be whereas this is clear to me in this approach.

**Comment by dalg24** [comment]
> > You forgot the deallocation
> 
> Can you please elaborate? `posix_memalign` allocations are also deallocated using `free`.

Just me being ignorant.  Sorry about the noise.

**Comment by dalg24** [comment]
> > Why did you chose this approach rather than the one suggested by Jonathan [#6367 (comment)](https://github.com/kokkos/kokkos/issues/6367#issuecomment-1683435438)
> 
> It's not that clear to me what the fallback(s) would be whereas this is clear to me in this approach.

Same fallbacks as you have now

**Comment by barracuda156** [comment]
Thank you for working to fix this!

**Comment by masterleinad** [comment]
@barracuda156 Ca you confirm that the latest version works for you?

**Review by crtrott — APPROVED** [review]
I like this, assuming it passes testing :-)

**Comment by masterleinad** [comment]
@msimberg Can you reproduce the test faiures for the HPX CI here?

**Comment by masterleinad** [comment]
> I like this, assuming it passes testing :-)

Not quite sure why `HPX` is complaining and if special casing for it would make sense.

**Comment by barracuda156** [comment]
@masterleinad Yes, from 183c05a3ca43b3830a5d6dcb28f375cbe85687a8 it builds fine for me. Thank you!

**Review by dalg24 — APPROVED** [review]

**Review by msimberg — APPROVED** [review]

**Comment by masterleinad** [comment]
`HIP-ROCm-5.2-C++20` is timing out. Everything else is passing.

File: core/src/impl/Kokkos_HostSpace.cpp [hunk 2]
```diff
@@ -46,6 +46,10 @@
 #include <sstream>
 #include <cstring>
 
+#ifdef KOKKOS_COMPILER_INTEL
+#include <aligned_new>
```
Comment (dalg24) [inline core/src/impl/Kokkos_HostSpace.cpp:50]: @masterleinad What was that header needed for?

Comment (barracuda156) [reply core/src/impl/Kokkos_HostSpace.cpp:50]: For Intel I have no idea. I only built this with GCC.

Comment (masterleinad) [reply core/src/impl/Kokkos_HostSpace.cpp:50]: Have a look at https://godbolt.org/z/jThT4h9Ed.

Comment (dalg24) [reply core/src/impl/Kokkos_HostSpace.cpp:50]: Understood.  Thanks.

**Pull request merged**